File: sach/de3368.py
//DE3368.PY - Tran Huu Nam - hsgtin.vn - 2024
import bisect
from collections import defaultdict
N = 10**6 + 1
n, m, h, w = 0, 0, 0, 0
id = []
event = defaultdict(list)
save = defaultdict(list)
a = []
B = []
b = []
def up(x, y, val):
    while x <= n:
        _y = y
        while _y <= m:
            B[x][_y] += val
            _y += (_y & -_y)
        x += (x & -x)
def get(x, y):
    ans = 0
    while x:
        _y = y
        while _y:
            ans += B[x][_y]
            _y -= (_y & -_y)
        x -= (x & -x)
    return ans
def sum(x, y, _x, _y):
    return get(_x, _y) - get(_x, y - 1) - get(x - 1, _y) + get(x - 1, y - 1)
def main():
    global n, m, h, w, a, B, b
    n, m, h, w = map(int, input().split())
    a = [[0] * (m + 1) for _ in range(n + 1)]
    B = [[0] * (m + 1) for _ in range(n + 1)]
    b = [[[0] * 20 for _ in range(m + 1)] for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            a[i][j] = int(input())
            id.append(a[i][j])
    id.sort()
    id = list(dict.fromkeys(id))
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            a[i][j] = bisect.bisect_left(id, a[i][j]) + 1
    for k in range(20):
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                b[i][j][k] = b[i - 1][j][k] + b[i][j - 1][k] - b[i - 1][j - 1][k] + ((a[i][j] >> k) & 1)
    ans = 0
    lim = h * w
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            event[a[i][j]].append((i, j))
            if i < h or j < w:
                continue
            mask = 0
            for k in range(20):
                cnt = b[i][j][k] - b[i - h][j][k] - b[i][j - w][k] + b[i - h][j - w][k]
                if cnt * 2 > lim:
                    mask |= (1 << k)
            save[mask].append((i, j))
    for i in range(1, N):
        if not save[i]:
            continue
        for x in event[i]:
            up(x[0], x[1], 1)
        for v in save[i]:
            x = v[0] - h + 1
            y = v[1] - w + 1
            _x = v[0]
            _y = v[1]
            if sum(x, y, _x, _y) * 2 > lim:
                ans += 1
        for x in event[i]:
            up(x[0], x[1], -1)
    print((n - h + 1) * (m - w + 1) - ans)
if __name__ == "__main__":
    main()

File: sach/de3368.cpp
//DE3368.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
#define f first
#define s second
#define all(x) x.begin(), x.end()
typedef long long ll;
typedef pair <int, int> ii;
const int N = 1e6 + 1;
int n, m, h, w;   vector <int> id;
vector <ii> event[N], save[N]; vector <vector <int>> a, B;
vector <vector <vector <int>>> b;
void up(int x, int y, int val) {
    for(; x <= n; x += (x & -x))
        for(int _y = y; _y <= m; _y += (_y & -_y))
            B[x][_y] += val;
}
int get(int x, int y) {
    int ans = 0;
    for(; x; x -= (x & -x))
        for(int _y = y; _y; _y -= (_y & -_y))
            ans += B[x][_y];
    return ans;
}
int sum(int x, int y, int _x, int _y) {
    return get(_x, _y)-get(_x,y-1)-get(x-1,_y)+get(x-1,y-1);
}
int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> n >> m >> h >> w;
    a.assign(n + 1, vector <int> (m + 1, 0));
    B.assign(n + 1, vector <int> (m + 1, 0));
    b.assign(n+1,vector<vector<int>>(m+1,vector<int>(20,0)));
    for(int i = 1; i <= n; i++)
        for(int j = 1; j <= m; j++) {
            cin >> a[i][j];
            id.push_back(a[i][j]);
        }
    sort(all(id));
    id.erase(unique(all(id)), id.end());
    for(int i = 1; i <= n; i++)
        for(int j = 1; j <= m; j++)
            a[i][j]=lower_bound(all(id),a[i][j])-id.begin()+1;
    for(int k = 0; k < 20; k++)
        for(int i = 1; i <= n; i++)
            for(int j = 1; j <= m; j++)
                b[i][j][k] = b[i-1][j][k] + b[i][j-1][k] –
                          b[i-1][j-1][k] + (a[i][j] >> k & 1);
    int ans = 0, lim = h * w;
    for(int i = 1; i <= n; i++) {
        for(int j = 1; j <= m; j++) {
            event[a[i][j]].push_back({i, j});
            if (i < h || j < w) continue;
            int mask = 0;
            for(int k = 0; k < 20; k++) {
                int cnt = b[i][j][k]-b[i-h][j][k]-b[i][j-w][k]
                             +b[i-h][j-w][k];
                if (cnt * 2 > lim)
                    mask |= (1 << k);
            }
            save[mask].push_back({i, j});
        }
    }
    for(int i = 1; i < N; i++) {
        if (save[i].size()==0) continue;
        for(auto x: event[i])
            up(x.f, x.s, 1);
        for(auto v: save[i]) {
            int x = v.f-h+1, y=v.s-w+1, _x=v.f, _y=v.s;
            if (sum(x, y, _x, _y) * 2 > lim) ans++;
        }
        for(auto x: event[i])
            up(x.f, x.s, -1);
    }
    cout << (n - h + 1) * (m - w + 1) - ans << "\n";
}

